### ISSUE TITLE
Refactor: models to support

### ISSUE THREAD

**Issue by sgreenbury** [alan-turing-institute/autoemulate#326]
Current models include:
- sklearn models (e.g. SVMs, radial basis functions, etc) - these support minimal code applications #320
- GPs/neural processes (already implemented with PyTorch)

Aim here from refactor to be "PyTorch-first" while enabling current users to continue using the refactored version.

Additionally, new models to include:
- Sparse GPs
- ...

**Comment by sgreenbury** [comment]
From discussion with @radka-j, some ideas for initial models to support:
- Gaussian Processes (GPyTorch Backend)
- Conditional Neural Processes (PyTorch Backend)
- Decision tree-based methods: aim for light GBM as widely used (`LightGBM`) (this can run on GPUs)
- If we have time, radial basis functions next

**Comment by radka-j** [comment]
I have some questions about whether we need all of the different tree based methods although there might be some pros/cons of each for different sized datasets. In order of complexity one could order them as:
- Random forests
- Gradient boosting from sklearn 
- LightGBM (i.e., gradient boosting but optimised for large datasets)
    - sklearn now has a [regressor](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.HistGradientBoostingRegressor.html#sklearn.ensemble.HistGradientBoostingRegressor) inspired by LightGBM that is also meant to perform well on large datasets

**Comment by sgreenbury** [comment]
Following discussion with the team, the current models are aiming for refactor and 1.0
- #545
- #451, specifically a MLP model in first instance given data structure currently supported.
  - Beyond 1.0/refactor: UQ extensions include looking at approaches involving adding uncertainty as an output head and with Bayesian last layer, BNNs, etc (#549)
- #429: adding support for ensembles would bring UQ to any of the models

Currently sklearn methods such as RandomForest are supported through the `SklearnBackend` - as discussed, post 1.0 it would be worth considering how this could be implemented to work directly with torch backend (potential advantage is parallelisation over trees on GPU, differentiability might be possible for some models).

Closing this issue with the above discussion capturing aims for the refactor.
